device-connectors/src/testflinger_device_connectors/devices/zapper/__init__.py

```diff
@@ -0,0 +1,102 @@
+# Copyright (C) 2024 Canonical
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Package containing modules for implementing Zapper-driven device connectors.
+
+Modules inheriting from the provided abstract class will run Zapper-driven
+provisioning procedures via Zapper API. The provisioning logic is implemented
+in the Zapper codebase and the connector serves as a pre-processing step,
+validating the configuration and preparing the API arguments.
+"""
+
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Tuple
+
+import rpyc
+import yaml
+
+import testflinger_device_connectors
+from testflinger_device_connectors.devices import (
+    DefaultDevice,
+    RecoveryError,
+    catch,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ZapperConnector(ABC, DefaultDevice):
+    """
+    Abstract base class defining a common interface for Zapper-driven
+    device connectors.
+    """
+
+    PROVISION_METHOD = ""  # to be defined in the implementation
+    ZAPPER_REQUEST_TIMEOUT = 60 * 90
+    ZAPPER_SERVICE_PORT = 60000
+
+    @catch(RecoveryError, 46)
+    def provision(self, args):
+        """Method called when the command is invoked."""
+        with open(args.config) as configfile:
+            config = yaml.safe_load(configfile)
+        testflinger_device_connectors.configure_logging(config)
+
+        (api_args, api_kwargs) = self._validate_configuration(
+            args.config, args.job_data
+        )
+
+        logger.info("BEGIN provision")
+        logger.info("Provisioning device")
+
+        self._run(args.config["controller_host"], *api_args, **api_kwargs)
+
+        logger.info("END provision")
+
+    @abstractmethod
+    def _validate_configuration(
+        self, config, job_data
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        """
+        Validate the job config and data and prepare the arguments
+        for the Zapper `provision` API.
+        """
```

Comment:
shall we raise NotImplementedError here as well?

Reply:
The method is abstract, you'll get a TypeError if you try to create an object out of this class.

Reply:
But what if a child class doesn’t implement this method?

Reply:
Same effect, it'll raise TypeError.

Reply:
Hmm, this is what's written in the [documentation](https://docs.python.org/3/library/abc.html#abc.abstractmethod):
> Note Unlike Java abstract methods, these abstract methods may have an implementation. This implementation can be called via the super() mechanism from the class that overrides it.

So, you can do something like this
![image](https://github.com/canonical/testflinger/assets/84857215/dcedf988-a774-4eac-8184-53209b35f1b7)

I believe the desired behaviour is to tell a user that the method is not implemented, so the code would return an error if the exception was raised

Reply:
In this case I don't want anything implemented in the abstract method (so `super()` is not relevant). I want the user to implement it. If they don't do it, they'll get a `TypeError` even before they try to run that method:

```
class A(ABC):
    @abstractmethod
    def mymethod(self):
        """Implement me."""
        
class B(A):
    ...
    
obj = B()  <-- this will raise TypeError
```

What I'm doing here is defining an interface so that every Zapper connector behave in the same way.

Reply:
but in my case I also don't have anything implemented in my abstract class, and you still can call `super()` there. That's why I suggest raising an error in the abstract method itself, because marking it with `abstractmethod` doesn't mean we cannot call it

Reply:
moreover, you can do something like this as well without seeing an error :D
 
```python3
>>> class A(ABC):
...     @abstractmethod
...     def mymethod(self):
...         """Implement me."""
... 
>>> A.mymethod(1)
```

Reply:
Abstract class and abstract methods are designed for this exact purpose: making the child implement what's abstract and eventually override the default methods. I don't see benefits in raising an exception is someone runs `super()`, I'm actually fine with that as long as they implement their own stuff. Moreover, that method expects a return value, so a mere `return super().method()` implementation wouldn't be enough.

And for what concerns that example: not true, `mymethod` is not a static/class method so no way to call it with `A.mymethod(1)`. You need to create an object first which will raise `TypeError`.

Reply:
You got it :) https://github.com/canonical/testflinger/pull/198/commits/a99f6ff643e5949a804e42b470c165b589415921